src/ucs/datastruct/arbiter.c

```diff
@@ -254,80 +292,112 @@ void ucs_arbiter_dispatch_nonempty(ucs_arbiter_t *arbiter, unsigned per_group,
         ucs_arbiter_group_head_reset(group_head);
 
         group_dispatch_count = 0;
-        sched_group          = 1;
         group                = group_head->group;
+        dummy.group          = group;
         UCS_ARBITER_GROUP_GUARD_CHECK(group);
 
-        do {
-            /* zero pointer to next elem here because:
+        for (;;) {
+            ucs_assert(group_head->group   == group);
+            ucs_assert(dummy.group         == group);
+            ucs_assert(group_dispatch_count < per_group);
+
+            /* reset the dispatched element here because:
              * 1. if the element is removed from the arbiter it must be kept in
              *    initialized state otherwise push will fail
-             * 2. we can't zero the pointer after calling the callback because
+             * 2. we can't reset the element after calling the callback because
              *    the callback could release the element.
              */
-            next_elem        = group_head->next;
-            group_head->next = NULL;
-            ucs_assert(group_head->group == group);
+            ucs_arbiter_elem_init(group_head);
+            ucs_assert(!ucs_arbiter_group_head_is_scheduled(group_head));
+
+            /* replace group head by a dummy element, to allow scheduling more
+             * elements on this group from the dispatch callback.
+             */
+            ucs_arbiter_group_head_replace(group, group_head, &dummy);
 
+            /* dispatch the element */
             ucs_trace_poll("dispatching arbiter element %p", group_head);
             UCS_ARBITER_GROUP_GUARD_ENTER(group);
             result = cb(arbiter, group, group_head, cb_arg);
             UCS_ARBITER_GROUP_GUARD_EXIT(group);
             ucs_trace_poll("dispatch result: %d", result);
             ++group_dispatch_count;
 
-            if (result == UCS_ARBITER_CB_RESULT_REMOVE_ELEM) {
-                group_tail = group->tail;
-                if (group_head == group_tail) {
-                    /* Last element */
-                    group->tail = NULL; /* Group is empty now */
-                    sched_group = 0;
-                    group_head  = NULL; /* for debugging */
+            /* should not push to head */
```

Comment:
why? we have api for this..

Reply:
only push to tail is allowed during dispatch, push to head is harder to support and not needed for UCT pending add for now